### ISSUE TITLE
webpacker 6.0.0.beta.5 removal of javascript_packs_with_chunks_tags

### ISSUE THREAD

**Issue by mmaloon** [rails/webpacker#2918]
In my app I use use a combination of the `javascript_packs_with_chunks_tags` and the `javascript_pack_tags` helpers.

Every page renders the entry points with the `javascript_packs_with_chunks_tags` helper, and then conditionally I would also  preload some of the non-entry chunks (that I know where required for the page based on the route) using the old `javascript_pack_tags`. 

```
    <%= javascript_packs_with_chunks_tag 'main-bundle' %>
    <% if @_preload_async_chunks.present? %>
      <%= javascript_pack_tag *@_preload_async_chunks %>
    <% end %>
```

With this code, those async chunks can be fetched once the page loads without having to wait for the entrypoint chunk to execute. Now that the ability to include non-entrypoint chunks has been removed, I cannot do this optimization.

**Comment by rossta** [comment]
I believe can still use script tags with asset_pack_path. But this sounds more like you want to use dynamic imports? That’s webpack’s sweet spot. You could use the gon gem to pass import names to JS from Ruby to async load selected modules.

**Comment by mmaloon** [comment]
Thanks @rossta. I actually am using dynamic imports extensively. The issue is that the async (dynamic) chunk will only get fetched by the client once that dynamic `import` statement within the entrypoint bundle is executed. This means the entry bundle must first be downloaded, compiled and processed before the async chunk will be fetched.
Your tip to use the `asset_pack_path` worked with a little modification:

```ruby
<%= javascript_pack_tag 'main-bundle' %>
<% if @_preload_async_chunks.present? %>
  <%= javascript_include_tag *@_preload_async_chunks.map{|p| asset_pack_path("#{p}.js")} %>
<% end %>
```

**Comment by feliperaul** [comment]
@mmaloon @rossta I'm still on Webpacker 5, but as far as I understand, webpacker 6 `javascript_include_tag` is the old `javascript_packs_with_chunks_tag`.

So this is how we're doing to handle the need described by @mmaloon . I'd love to listen to @rossta 's opinion regarding this approach. 

1. In your ApplicationHelper, define a memoized 'accumulator' for the additional entry packs like this:

```
def additional_webpacker_entry_packs_to_load
    @additional_webpacker_entry_packs_to_load ||= Set.new
  end
```

2. In your layouts, change your `javascript_include_tag` (or `javascript_packs_with_chunks_tag` if you are using Webpacker 4 or 5) to this (we're splatting that set):

```
<%= javascript_packs_with_chunks_tag "application", *additional_webpacker_entry_packs_to_load %>
```

3. In your views or partials, whenever you need to load additional entry packs specific to them, you simply put this anywhere (there's no need for `content_for`, just drop it anywhere):

```
<% additional_webpacker_entry_packs_to_load.add('conversations') %>
```

This gives us maximum flexibility. 


We still use *dynamic imports* on many pages, but only for javascript that installs behavior *below the fold*, because as @mmaloon pointed out, dynamic imports is not ideal for pages that need additional javascript for behavior *above the fold*, because for these to have maximum performance it's good to load them as soon as possible, not having to wait for the `packs/application.js` to be loaded, parsed, and then fire the extra request for the needed dynamic import.

So, in a nutshell, this is how we're doing it:

- If you need page-specific javascript that will install behavior below the fold, or it's not essential for time to interactive, you're better off using the dynamic import (`import()`) inside a `DOMContentLoaded` event listener;
- If you need page-specific javascript that will install behavior above the fold, create a separate entry pack in `packs/` for it, and then use `additional_webpacker_entry_packs_to_load` in the view to make sure it will be loaded alongside the other entry packs with common chunks without any dupplication.

What do you guys think? Maybe @rossta can even write in his amazing blog about this :D
